File: labs/10-mcp-client/python/main.py
"""Minimal MCP client over stdio, plus an SSE parsing demonstration.

Spawns the server from Lab 09 as a subprocess, performs the initialize
handshake, lists tools, calls one, then demonstrates how SSE events are parsed.
"""

import json
import logging
import os
import subprocess
import sys
import time
from pathlib import Path
from typing import Any, Iterator

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import configure_logging

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
logger = logging.getLogger("mcp-client")


def _default_server_script() -> Path:
    return Path(__file__).resolve().parents[2] / "09-mcp-server-stdio" / "python" / "main.py"


class MCPStdioClient:
    """A tiny MCP client that drives a server subprocess over stdio."""

    def __init__(self, server_script: str | Path | None = None) -> None:
        script = Path(server_script or os.getenv("MCP_SERVER_SCRIPT") or _default_server_script())
        if not script.exists():
            raise FileNotFoundError(f"MCP server script not found: {script}")
        self.script = script
        self.process: subprocess.Popen | None = None
        self._request_id = 0

    def _next_id(self) -> int:
        self._request_id += 1
        return self._request_id

    def connect(self) -> None:
        logger.info("Starting MCP server: %s", self.script)
        self.process = subprocess.Popen(
            [sys.executable, str(self.script)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=False,
            bufsize=0,
        )

    def disconnect(self) -> None:
        if self.process and self.process.poll() is None:
            self.process.stdin.close()
            self.process.terminate()
            try:
                self.process.wait(timeout=2)
            except subprocess.TimeoutExpired:
                self.process.kill()
        self.process = None

    def _send(self, message: dict[str, Any]) -> None:
        body = json.dumps(message, ensure_ascii=False).encode("utf-8")
        data = f"Content-Length: {len(body)}\r\n\r\n".encode("utf-8") + body
        self.process.stdin.write(data)
        self.process.stdin.flush()

    def _recv(self) -> dict[str, Any]:
        headers: dict[str, str] = {}
        while True:
            line = self.process.stdout.readline()
            if not line:
                raise ConnectionError("Server closed stdout")
            line = line.decode("utf-8").strip()
            if not line:
                break
            if ":" in line:
                key, value = line.split(":", 1)
                headers[key.strip().lower()] = value.strip()

        length = int(headers.get("content-length", 0))
        if length == 0:
            raise ConnectionError("Empty message body")
        body = self.process.stdout.read(length)
        return json.loads(body.decode("utf-8"))

    def initialize(self) -> dict[str, Any]:
        self._send(
            {
                "jsonrpc": "2.0",
                "id": self._next_id(),
                "method": "initialize",
                "params": {"protocolVersion": "2024-11-05", "capabilities": {}},
            }
        )
        result = self._recv()
        self._send(
            {"jsonrpc": "2.0", "id": None, "method": "notifications/initialized"}
        )
        return result

    def list_tools(self) -> list[dict[str, Any]]:
        self._send({"jsonrpc": "2.0", "id": self._next_id(), "method": "tools/list"})
        response = self._recv()
        if "error" in response:
            raise RuntimeError(response["error"])
        return response.get("result", {}).get("tools", [])

    def call_tool(self, name: str, arguments: dict[str, Any]) -> dict[str, Any]:
        self._send(
            {
                "jsonrpc": "2.0",
                "id": self._next_id(),
                "method": "tools/call",
                "params": {"name": name, "arguments": arguments},
            }
        )
        response = self._recv()
        if "error" in response:
            raise RuntimeError(response["error"])
        return response.get("result", {})


def parse_sse(lines: Iterator[bytes]) -> Iterator[dict[str, str]]:
    """Parse Server-Sent Events from an iterable of byte lines.

    Example SSE stream:
        event: message
        data: {"id": 1, "text": "hello"}

        event: done
        data: EOF
    """
    event: dict[str, str] = {}
    for raw in lines:
        line = raw.decode("utf-8").rstrip("\n").rstrip("\r")
        if not line:
            if event:
                yield event
                event = {}
            continue
        if line.startswith(":"):
            continue  # comment
        if ":" in line:
            key, value = line.split(":", 1)
            value = value.lstrip()
            event[key] = value
    if event:
        yield event


def demo_sse() -> None:
    """Demonstrate SSE parsing with a fake byte stream."""
    raw_stream = (
        b":heartbeat\n\n"
        b"event: message\n"
        b"data: {\"tool\": \"calculate\", \"args\": {\"expression\": \"1+1\"}}\n\n"
        b"event: status\n"
        b"data: processing\n\n"
        b"event: done\n"
        b"data: finished\n\n"
    )
    print("\n[SSE transport concept]")
    for event in parse_sse(iter(raw_stream.splitlines(keepends=True))):
        print("  SSE event:", event)


def main():
    configure_logging()
    client = MCPStdioClient()
    try:
        client.connect()
        init_response = client.initialize()
        print("[initialize]", json.dumps(init_response.get("result", {}), indent=2))

        tools = client.list_tools()
        print("\n[tools]")
        for tool in tools:
            print(f"  - {tool['name']}: {tool.get('description', '')}")

        result = client.call_tool("calculate", {"expression": "(10 + 5) / 3"})
        print("\n[tools/call calculate]")
        for item in result.get("content", []):
            print(" ", item.get("text"))

        demo_sse()
    except Exception as exc:
        logger.error("Client failed: %s", exc)
        raise
    finally:
        client.disconnect()


if __name__ == "__main__":
    main()

File: labs/10-mcp-client/node/main.ts
import { spawn, ChildProcessWithoutNullStreams } from "child_process";
import path from "path";

function defaultServerScript(): string {
  return path.resolve(__dirname, "..", "..", "09-mcp-server-stdio", "python", "main.py");
}

class StreamReader {
  private buffer = Buffer.alloc(0);
  private waiter: (() => void) | null = null;
  private ended = false;

  constructor(stream: NodeJS.ReadableStream) {
    stream.on("data", (chunk: Buffer) => {
      this.buffer = Buffer.concat([this.buffer, chunk]);
      if (this.waiter) {
        this.waiter();
        this.waiter = null;
      }
    });
    stream.on("end", () => {
      this.ended = true;
      if (this.waiter) {
        this.waiter();
        this.waiter = null;
      }
    });
  }

  private async wait(): Promise<void> {
    if (this.buffer.length > 0 || this.ended) return;
    await new Promise<void>((resolve) => {
      this.waiter = resolve;
    });
  }

  async readUntil(sep: string): Promise<Buffer | null> {
    const sepBuf = Buffer.from(sep);
    while (true) {
      const idx = this.buffer.indexOf(sepBuf);
      if (idx !== -1) {
        const result = this.buffer.slice(0, idx);
        this.buffer = this.buffer.slice(idx + sepBuf.length);
        return result;
      }
      if (this.ended) return this.buffer.length > 0 ? this.buffer : null;
      await this.wait();
    }
  }

  async readExactly(n: number): Promise<Buffer | null> {
    while (this.buffer.length < n) {
      if (this.ended) return null;
      await this.wait();
    }
    const result = this.buffer.slice(0, n);
    this.buffer = this.buffer.slice(n);
    return result;
  }
}

class MCPStdioClient {
  private process: ChildProcessWithoutNullStreams | null = null;
  private requestId = 0;
  private reader: StreamReader | null = null;

  constructor(private serverScript?: string) {}

  private nextId(): number {
    return ++this.requestId;
  }

  connect(): void {
    const script =
      this.serverScript || process.env.MCP_SERVER_SCRIPT || defaultServerScript();
    console.log("Starting MCP server:", script);
    this.process = spawn("python", [script], {
      stdio: ["pipe", "pipe", "pipe"],
    });
    this.reader = new StreamReader(this.process.stdout);
  }

  disconnect(): void {
    if (this.process && !this.process.killed) {
      this.process.stdin.end();
      this.process.kill();
    }
    this.process = null;
  }

  private send(message: any): void {
    const body = Buffer.from(JSON.stringify(message), "utf8");
    const data = Buffer.concat([
      Buffer.from(`Content-Length: ${body.length}\r\n\r\n`),
      body,
    ]);
    this.process!.stdin.write(data);
  }

  private async recv(): Promise<any> {
    const headerBytes = await this.reader!.readUntil("\r\n\r\n");
    if (!headerBytes) throw new Error("Server closed stdout");
    const headers: Record<string, string> = {};
    for (const line of headerBytes.toString("utf8").split("\r\n")) {
      const colon = line.indexOf(":");
      if (colon !== -1) {
        headers[line.slice(0, colon).trim().toLowerCase()] = line
          .slice(colon + 1)
          .trim();
      }
    }
    const length = parseInt(headers["content-length"] || "0", 10);
    if (length === 0) throw new Error("Empty message body");
    const body = await this.reader!.readExactly(length);
    if (!body) throw new Error("Server closed stdout while reading body");
    return JSON.parse(body.toString("utf8"));
  }

  async initialize(): Promise<any> {
    this.send({
      jsonrpc: "2.0",
      id: this.nextId(),
      method: "initialize",
      params: { protocolVersion: "2024-11-05", capabilities: {} },
    });
    const result = await this.recv();
    this.send({
      jsonrpc: "2.0",
      id: null,
      method: "notifications/initialized",
    });
    return result;
  }

  async listTools(): Promise<any[]> {
    this.send({ jsonrpc: "2.0", id: this.nextId(), method: "tools/list" });
    const response = await this.recv();
    if (response.error) throw new Error(JSON.stringify(response.error));
    return response.result?.tools || [];
  }

  async callTool(name: string, args: any): Promise<any> {
    this.send({
      jsonrpc: "2.0",
      id: this.nextId(),
      method: "tools/call",
      params: { name, arguments: args },
    });
    const response = await this.recv();
    if (response.error) throw new Error(JSON.stringify(response.error));
    return response.result || {};
  }
}

function parseSSE(lines: string[]): Array<Record<string, string>> {
  const events: Array<Record<string, string>> = [];
  let event: Record<string, string> = {};
  for (const raw of lines) {
    const line = raw.replace(/\r$/, "");
    if (!line) {
      if (Object.keys(event).length > 0) {
        events.push(event);
        event = {};
      }
      continue;
    }
    if (line.startsWith(":")) continue;
    const colon = line.indexOf(":");
    if (colon !== -1) {
      const key = line.slice(0, colon);
      const value = line.slice(colon + 1).replace(/^\s+/, "");
      event[key] = value;
    }
  }
  if (Object.keys(event).length > 0) events.push(event);
  return events;
}

function demoSSE(): void {
  const rawStream = Buffer.from(
    ":heartbeat\n\n" +
      "event: message\n" +
      'data: {"tool": "calculate", "args": {"expression": "1+1"}}\n\n' +
      "event: status\n" +
      "data: processing\n\n" +
      "event: done\n" +
      "data: finished\n\n"
  );
  console.log("\n[SSE transport concept]");
  for (const event of parseSSE(rawStream.toString("utf8").split(/\r?\n/))) {
    console.log("  SSE event:", event);
  }
}

async function main() {
  const client = new MCPStdioClient();
  try {
    client.connect();
    const initResponse = await client.initialize();
    console.log("[initialize]", JSON.stringify(initResponse.result || {}, null, 2));

    const tools = await client.listTools();
    console.log("\n[tools]");
    for (const tool of tools) {
      console.log(`  - ${tool.name}: ${tool.description || ""}`);
    }

    const result = await client.callTool("calculate", { expression: "(10 + 5) / 3" });
    console.log("\n[tools/call calculate]");
    for (const item of result.content || []) {
      console.log(" ", item.text);
    }

    demoSSE();
  } catch (exc: any) {
    console.error("Client failed:", exc);
    process.exit(1);
  } finally {
    client.disconnect();
  }
}

main();
